### DISCUSSION TITLE
Support for WPF in current version?

### DISCUSSION THREAD

**Discussion by Awsmolak** [dotnet/interactive#3148] [Q&A] [answered]
I want to be able to call methods from a project that targets TFM ```.net7.0-windows```.

All goes well until I run into any dependencies on WindowsBase. In this case System.Windows.Point. 
```System.TypeLoadException: Could not load type 'System.Windows.Size' from assembly 'WindowsBase, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.```

I've tried various methods trying to reference or install WindowsBase in my notebook without success. Although I was surprised that I got as far as I did. I am assuming that .net interactive just doesn't support .Net core in its entirety. 

I decided to try the approach with an embedded C# kernel running in my app and connecting to it through named pipes. However, the [wpf-connect](https://github.com/dotnet/interactive/tree/main/samples/connect-wpf) example code does not work with the current release. (See #3146 ) 

I'm looking for any suggestions using either of the two approaches I'm trying to take. 

My next step will be digging into the code and docs to try and figure out how to update the code in my app (and the example) to work with the current version of the pipes API. But I'm running down the clock for how much time I have to evaluate tooling for the project I want to use this on. So any help towards the most efficient path forward would be appreciated!

**Comment by jonsequitur** [comment]
I think this issue is because .NET Interactive is not a .NET Framework application. It currently targets .NET 7. A proxy kernel would need to be loaded if you want to host it in a .NET Framework application. 

#3144 could help to address this.

I'll take a look at the WPF sample.

**Reply by Awsmolak** [reply]
I did update my app to multitarget .Net 7 and I tested it standalone, all the relevant pieces run and dependencies load in the .7 runtime. I made sure I was importing the correct dll in the notebook (```#r "path to my net7-windows build/Myapp.dll"```). (Due to circumstances beyond our control our production app needs to be build on .net framework, hence the multi-targeting)

Thanks for looking into this. It has really really good potential!

As soon as you update the sample project to use the latest .NET interactive release, you'll see that ```CompositeKernel.UseLogMagicCommand()``` is gone and ```CsharpKernel.SetValueAsync("App", this)``` should probably be ```CsharpKernel.SetValueAsync("App", this, typeof(App))```. So, maybe it's an easy fix.

**Reply by jonsequitur** [reply]
Thanks for the pointer. I also expect it's an easy fix.

**Reply by jonsequitur** [reply]
#3149.

**Reply by Awsmolak** [reply, marked as answer]
After updating I am able to connect to my app and execute using pipes!

I'm also able to execute the code by importing the relevant .dlls. I just had to reference the correct WindowsBase: 
`#r "C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App\7.0.10\WindowsBase.dll"` 

Also, I suggest in [App.xaml.cs](https://github.com/dotnet/interactive/blob/main/samples/connect-wpf/App.xaml.cs) the line:
```csharp
//Load WPF app assembly 
await csharpKernel.SendAsync(new SubmitCode(@$"#r ""{typeof(App).Assembly.Location}""
using {nameof(WpfConnect)};"));
```
be changed to
```csharp
//Load WPF app assembly 
await csharpKernel.SendAsync(new SubmitCode(@$"#r ""{typeof(App).Assembly.Location}""
using {System.Reflection.Assembly.GetExecutingAssembly().GetName().Name};"))
```
That will allow it to work with an assemblies that have nested namespaces. Took me a minute to figure out why it wasn't working for me.
